Feature/420

TColl opened the pull request:
Relates to #420 

Here's a first attempt at parsing frictionless schema, with the following capabilities:

- read from frictionless schema json/yaml files on disk
- accept a frictionless Schema object directly from memory
- map field names, types, constraints directly into pandera equivalents

The code to handle this is in `pandera.io`, and I've created a new `FrictionlessFieldParser` class to convert each field in the schema accordingly.

Currently `pa.io.from_frictionless_schema()` is as far as I've got in terms of an API - how/where do you want this to be imported from?

Documentation will also need updating - there are docstrings for everything I've added but nothing more than that at this point.

I've also added tests in `tests.io` to validate this functionality.

TColl commented:
@cosmicBboy I've swapped this over to target v0.7.0 and added an example to the docstring so documentation should be a bit clearer now.

I've just realised I branched from master rather than from 0.7.0 directly, sorry - do you want this fixing so only my commits get covered by this MR?

Please let me know if you need anything else from me! I can possibly take a look at parsing open API schema too, though I haven't used them before. It should be easy to refactor what I've done here for anyone to extend coverage to other schema quite consistently in future, too (e.g. an abstract base class for column parsing, which is extended by `FrictionlessField`, `OpenAPIColumn` etc with a consistent API for constructing a pandera schema from them).

cosmicBboy commented:
thanks @TColl this looks fantastic! ✨

> I've just realised I branched from master rather than from 0.7.0 directly, sorry - do you want this fixing so only my commits get covered by this MR?

No worries, the branch situation is still evolving as more folks start contributing to the project! I just updated the `release/0.7.0` branch to be on par with `master`. Would you mind rebasing your changes on top of the latest `release/0.7.0` branch and re-opening the PR?

Will add comments/feedback after the rebase

TColl commented:
> thanks @TColl this looks fantastic! ✨
> 
> > I've just realised I branched from master rather than from 0.7.0 directly, sorry - do you want this fixing so only my commits get covered by this MR?
> 
> No worries, the branch situation is still evolving as more folks start contributing to the project! I just updated the `release/0.7.0` branch to be on par with `master`. Would you mind rebasing your changes on top of the latest `release/0.7.0` branch and re-opening the PR?
> 
> Will add comments/feedback after the rebase

I've rebased now, though it's not pretty!

cosmicBboy commented:
huh, weird... can you try making a new PR off of this branch? I think you'll have to make a new branch in your fork https://github.com/TColl/pandera/commit/a98aeb92edda632fb2165d1c855884f8f75a2a77

it looks like https://github.com/TColl/pandera/tree/feature/420 doubles up  on the `improve documentation` and `parse frictionless schema ` commits since I made those changes to github actions file.

TColl commented:
> huh, weird... can you try making a new PR off of this branch? I think you'll have to make a new branch name from this commit in your fork [TColl@a98aeb9](https://github.com/TColl/pandera/commit/a98aeb92edda632fb2165d1c855884f8f75a2a77)
> 
> it looks like https://github.com/TColl/pandera/tree/feature/420 doubles up on the `improve documentation` and `parse frictionless schema ` commits since I made those changes to github actions file.

I've force-pished a clean branch over the old one and cherry-picked the changes back in from the messy one - should be good to review now 👍

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+
+class FrictionlessFieldParser:
+    """Parses frictionless data schema field specifications so we can convert
+    them to an equivalent pandera Column schema.
```
```suggestion
    them to an equivalent :class:`pandera.schema_components.Column` schema.
```

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+    them to an equivalent pandera Column schema.
+
+    For this implementation, we are using field names, constraints and types
+    but leaving other frictionless parameters out (e.g. foreign keys, type
```
we can go back and add this feature when #331 is done

TColl replied:
in anticipation of this, I've standardised all properties of `FrictionlessFieldParser` to reflect the properties we need for a pandera Column object - so when #331 is complete it'll be clearer where to add the new properties in, too.

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+        self.type = field.get("type", "string")
+
+    @property
+    def dtype(self):
```
@jeffzi just to give you a heads up, this might be one area we should revisit as your work with #369 is merged

cosmicBboy replied:
```suggestion
    def dtype(self) -> str:
```

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+        `here <https://specs.frictionlessdata.io/table-schema/#constraints>`_
+        and maps them into the equivalent pandera checks.
+
+        :returns: a list of pandera checks which capture the standard constraint
```
```suggestion
        :returns: a list of pandera :class:`pandera.checks.Check` objects which capture the standard constraint
```

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+        )
+
+    @property
+    def checks(self):
```
```suggestion
    def checks(self) -> Optional[List[Check]]:
```

TColl replied:
this actually returns a dict which is deserialized into a list of checks later - now updated and docstring is clearer on this too

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+            """Catches bounded constraints where we need to combine a min and max
+            pair of constraints into a single check."""
+            if (
+                min_constraint in constraints.keys()
```
consider simplifying to
```python
if min_constraint in constraints and max_constraint in constraints:
```

since `x in dict` checks for presence of x in keys

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+        }
+
+
+def from_frictionless_schema(schema):
```
```suggestion
def from_frictionless_schema(schema: FrictionlessSchema):
```

cosmicBboy replied:
oh, I see this doesn't necessarily have to be a `FrictionlessSchema`, feel free to update type here to allowable types (I'm not a frictionless schema user 😅)

cosmicBboy commented on pandera/io.py:
```diff
@@ -404,3 +405,196 @@ def to_script(dataframe_schema, path_or_buf=None):
+    column specification using :class:`~pandera.io.FrictionlessFieldParser`
+    to map field characteristics to pandera column specifications.
+
+    :param frictionless_schema: the frictionless schema object (or a
```
nit: make docstring consistent with argument name (either this or rename arg name to `frictionless_schema`

```suggestion
    :param schema: the frictionless schema object (or a
```

cosmicBboy reviewed (commented):
great work @TColl ! I added a few suggestions and nit, see inline comments

TColl commented:
@cosmicBboy after making these latest changes I'm getting pylint E1136 (unscriptable object) errors for the newly-used `typing.Optional` and `typing.Union` type hints with the current pre-commit setup - I've raised #455 to cover this

cosmicBboy commented:
hi @TColl I think for now you can put a `#pylint: disable=unsubscriptable-object` comment above the places where you get this error, this is accounted for in 3.9 tests [here](https://github.com/pandera-dev/pandera/blob/master/noxfile.py#L261).

I think #455 would be a good time to look at other, less opinionated linting options... I've been finding that pylint has become more and more of a burden, though maybe that's because of python's growing pains adding more features in recent versions. If we do decide to stick with pylint unpinning it will probably yield more issues, so we can localize those changes in #455.

It looks like there are also some punit test issues](https://github.com/pandera-dev/pandera/pull/454/checks?check_run_id=2267711686), are you able to reproduce them locally? See [here](https://pandera.readthedocs.io/en/stable/CONTRIBUTING.html#run-the-test-suite-locally) to run the testing suite that's on par with what CI runs.

TColl commented:
> hi @TColl I think for now you can put a `#pylint: disable=unsubscriptable-object` comment above the places where you get this error, this is accounted for in 3.9 tests [here](https://github.com/pandera-dev/pandera/blob/master/noxfile.py#L261).

I've resorted to running pre-commit once to check those are the only errors I get, then committing with a `--no-verify` flag to bypass pylint failing on those specific issues. Neither option is great, but adding pylint pragma always feels like a failure to me!

> I think #455 would be a good time to look at other, less opinionated linting options... I've been finding that pylint has become more and more of a burden, though maybe that's because of python's growing pains adding more features in recent versions. If we do decide to stick with pylint unpinning it will probably yield more issues, so we can localize those changes in #455.

I'm a fan of pylint personally, but I think #455 is a good place to have that discussion!
 
> It looks like there are also some [unit test issues](https://github.com/pandera-dev/pandera/pull/454/checks?check_run_id=2267711686), are you able to reproduce them locally? See [here](https://pandera.readthedocs.io/en/stable/CONTRIBUTING.html#run-the-test-suite-locally) to run the testing suite that's on par with what CI runs.

I get the same result locally - some of them are trivial, if annoying (`nan != nan` for some versions of python), and can be fixed with some updating of the unit tests themselves to be more robust. Others are more concerning to me (the `NoneType object has no 'get' method` ones) and will need a code tweak or two to get it sorted. I'll get this done in the next couple of days. 

On a side note, I'm a big fan of #397 purely for it halving the number of nox cases to run through - that full suite takes a while to burn through!

TColl commented:
@cosmicBboy sorry for the delay (I got a puppy, which has taken up most of my spare time!) but the unit tests are now fixed.This was mostly my unit testing not capturing various nuances between pandas versions so the actual code is largely unchanged.

cosmicBboy reviewed (approved):
> sorry for the delay (I got a puppy, which has taken up most of my spare time!)

nice! 🐶

thanks for the PR 🙏, and congrats to your first contribution to pandera 🎉! This is also the first of many integrations with other schema formats, so thanks for kicking that off 🍾
